typing/typecore.ml

```diff
@@ -3163,49 +3231,40 @@ and type_expect_
         exp_type = body.exp_type;
         exp_attributes = sexp.pexp_attributes;
         exp_env = env }
-  | Pexp_fun (l, Some default, spat, sbody) ->
-      assert(is_optional l); (* default allowed only with optional argument *)
-      let open Ast_helper in
-      let default_loc = default.pexp_loc in
-      let default_ghost = {default.pexp_loc with loc_ghost = true} in
-      let scases = [
-        Exp.case
-          (Pat.construct ~loc:default_ghost
-             (mknoloc (Longident.(Ldot (Lident "*predef*", "Some"))))
-             (Some ([], Pat.var ~loc:default_ghost (mknoloc "*sth*"))))
-          (Exp.ident ~loc:default_ghost (mknoloc (Longident.Lident "*sth*")));
-
-        Exp.case
-          (Pat.construct ~loc:default_loc
-             (mknoloc (Longident.(Ldot (Lident "*predef*", "None"))))
-             None)
-          default;
-       ]
-      in
-      let sloc =
-        { Location.loc_start = spat.ppat_loc.Location.loc_start;
-          loc_end = default_loc.Location.loc_end;
-          loc_ghost = true }
-      in
-      let smatch =
-        Exp.match_ ~loc:sloc
-          (Exp.ident ~loc (mknoloc (Longident.Lident "*opt*")))
-          scases
+  | Pexp_function (params, body_constraint, body) ->
+      let in_function = ty_expected_explained, loc in
+      let exp_type, params, body, newtypes =
+        type_function env params body_constraint body ty_expected ~in_function
+          ~first:true
       in
-      let pat = Pat.var ~loc:sloc (mknoloc "*opt*") in
-      let body =
-        Exp.let_ ~loc Nonrecursive
-          ~attrs:[Attr.mk (mknoloc "#default") (PStr [])]
-          [Vb.mk spat smatch] sbody
+      (* Require that the n-ary function is known to have at least n arrows
+         in the type. This prevents GADT equations introduced by the parameters
+         from hiding arrows from the resulting type.
+      *)
+      let ty_function =
+        List.fold_right
+          (fun param rest_ty ->
+             newty (Tarrow (param.fp_arg_label, newvar (), rest_ty, commu_ok)))
+          params
+          (match body with
+           | Tfunction_body _ -> newvar ()
+           | Tfunction_cases _ ->
+             newty (Tarrow (Nolabel, newvar (), newvar (), commu_ok)))
       in
-      type_function ?in_function loc sexp.pexp_attributes env
-                    ty_expected_explained l [Exp.case pat body]
-  | Pexp_fun (l, None, spat, sbody) ->
-      type_function ?in_function loc sexp.pexp_attributes env
-                    ty_expected_explained l [Ast_helper.Exp.case spat sbody]
-  | Pexp_function caselist ->
-      type_function ?in_function
-        loc sexp.pexp_attributes env ty_expected_explained Nolabel caselist
+      begin
+        try unify env ty_function exp_type
```

Comment:
If we did something like this, I think we'd want the criteria to be "whether any pattern contains a GADT" rather than whether patterns are active or not. Given the typability of functions may now have this additional requirement of unifying with an *n*-ary arrow type, it would be odd to enforce this requirement only in a hard-to-specify set of circumstances that depends on the presence/absence of active patterns.

Luckily, at time of writing, any function whose typability is affected by this additional requirement has to contain a GADT pattern in one of its parameters.

I don't think it's hard to check whether any pattern contains a GADT. (We already compute this for function parameters.) But it might be fragile, if in the future there are ways other than GADT patterns of running afoul of the new requirement.

My personal feeling is doing the check for GADTs is worth it. I'll go ahead and do that if you agree.

Reply:
We discussed this offline, and think doing the check just for GADTs (if it's easy) is a good way forward.

Reply:
This is now implemented.